recipes-openxt/qemu-dm/files/0027-nic-link-state-propagation.patch

```diff
@@ -0,0 +1,246 @@
+SHORT DESCRIPTION: 
+################################################################################
+Have the emulated NICs mirror the link-status of the network back-end, exposed
+through XenStore.
+
+################################################################################
+LONG DESCRIPTION: 
+################################################################################
+QEMU's current default network configuration is to have two NICs (rtl8139
+emulations), one for the Ethernet interface and one for the Wireless.
+Each emulated card is connected to a "QEMU vlan" (not 802.1q VLAN), actually an
+emulated hub (-net option behaviour), to which is also connected the tap
+interface of its backend.
+
+For each NIC created by QEMU, we add a XenStore watch on the node of the
+network back-end plugged in the same hub. This let us retrieve the back-end
+information using the nd_table (NICInfo).
+
+################################################################################
+CHANGELOG 
+################################################################################
+Port & documentation: Eric Chanudet, chanudete@ainfosec.com, 17/04/2015
+Intial Commit: Unknown
+
+################################################################################
+REMOVAL 
+################################################################################
+With the current OpenXT guest network configuration, removing this patch will
+trigger routing issues in the guest.
+
+################################################################################
+UPSTREAM PLAN 
+################################################################################
+There is no plan to upstream this patch, it is OpenXT specific.
+
+################################################################################
+INTERNAL DEPENDENCIES 
+################################################################################
+None
+
+################################################################################
+PATCHES 
+################################################################################
+--- a/hw/xen.h
++++ b/hw/xen.h
+@@ -38,6 +38,9 @@ void xen_piix_pci_write_config_client(ui
+ void xen_hvm_inject_msi(uint64_t addr, uint32_t data);
+ void xen_cmos_set_s3_resume(void *opaque, int irq, int level);
+ 
++int xenstore_register_nic(NetClientState *nc);
++int xenstore_unregister_nic(NetClientState *nc);
++
+ qemu_irq *xen_interrupt_controller_init(void);
+ 
+ int xen_init(void);
+--- a/hw/xen_backend.c
++++ b/hw/xen_backend.c
+@@ -38,6 +38,7 @@
+ #include "char/char.h"
+ #include "qemu/log.h"
+ #include "xen_backend.h"
++#include "qmp-commands.h"
+ 
+ #include <xen/grant_table.h>
+ 
+@@ -589,6 +590,143 @@ static int xenstore_scan(const char *typ
+     return 0;
+ }
+ 
++/*
++ * OpenXT: Get type and id from \([a-zA-Z]\+\)\([0-9]\+\).
++ */
++static void xenstore_nic_parse_name(const char *name, char *type, char *id)
++{
++    if (name) {
++        while (isalpha(*name)) {
```

Comment:
Output arguments, although we could add length argument if you think that is required.